tools/sigclient/pkg/ingest/ingest.go

```diff
@@ -151,10 +152,60 @@ func generateOpenTSDBBody(recs int, rdr utils.Generator) ([]byte, error) {
 	return retVal, nil
 }
 
-func runIngestion(iType IngestType, rdr utils.Generator, wg *sync.WaitGroup, url string, totalEvents int,
-	continous bool, batchSize, processNo int, indexPrefix string, ctr *uint64, bearerToken string,
-	indexName string, numIndices int) {
+var preGeneratedSeries []map[string]interface{}
 
+func generateUniqueID(series map[string]interface{}) string {
+	metricName := series["metric"].(string)
+	tags := series["tags"].(map[string]interface{})
+	var builder strings.Builder
+	builder.WriteString(metricName)
+	for key, value := range tags {
+		builder.WriteString(key)
+		builder.WriteString(fmt.Sprintf("%v", value))
+	}
+	return builder.String()
+}
+
+func generatePredefinedSeries(nMetrics int, cardinality int, gentype string) error {
+	preGeneratedSeries = make([]map[string]interface{}, 0, cardinality)
+	rdr, err := utils.InitMetricsGenerator(nMetrics, gentype)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	uniqueSeriesMap := make(map[string]struct{})
+	for len(preGeneratedSeries) < cardinality {
+		currPayload, err := rdr.GetRawLog()
+		if err != nil {
+			return err
+		}
+		uniqueID := generateUniqueID(currPayload)
+		if _, exists := uniqueSeriesMap[uniqueID]; !exists {
+			preGeneratedSeries = append(preGeneratedSeries, currPayload)
+			uniqueSeriesMap[uniqueID] = struct{}{}
+		}
+	}
+	elapsed := time.Since(start)
+	log.Infof("Generated %d unique series in %v", len(uniqueSeriesMap), elapsed)
+	return nil
+}
+
+func generateBodyFromPredefinedSeries(recs int) ([]byte, error) {
+	finalPayLoad := make([]interface{}, recs)
+	for i := 0; i < recs; i++ {
+		series := preGeneratedSeries[i%len(preGeneratedSeries)]
```

Comment:
you should not use the `i` for both for loop variable and to find the seriesId. Rather this series id var should be one continuous var, that keeps track of which series have been used so far